templates/components/engines/typescript/agent/tools/openapi-action.ts

```diff
@@ -0,0 +1,182 @@
+import SwaggerParser from "@apidevtools/swagger-parser";
+import axios, { AxiosResponse } from "axios";
+import { FunctionTool } from "llamaindex";
+
+interface DomainHeaders {
+  [key: string]: { [header: string]: string };
+}
+
+export class OpenAPIActionToolSpec {
+  private readonly INVALID_URL_PROMPT =
+    "This url did not include a hostname or scheme. Please determine the complete URL and try again.";
+
+  private readonly LOAD_OPENAPI_SPEC = {
+    name: "load_openapi_spec",
+    description:
+      "Use this function to load defined spec first before making requests.",
+    parameters: {
+      type: "object",
+      properties: {
+        url: {
```

Comment:
I guess this only supports loading from a URL for now, right? In Python, it can load from both a URL and a local file path (we also have the tool message to say that), so if we intend to load from a URL only, then we should change the example message in the tool config. https://github.com/run-llama/create-llama/blob/f7b14a7d161c94ac5c8329a24a56bb249ec30cfc/helpers/tools.ts#L138